src/redux/slices/userSlice.ts

```diff
@@ -0,0 +1,48 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+import { Graph, Profile } from "../../utilities/types";
+
+interface userState {
+  profile: Profile | null;
+  graph: Graph | null;
+}
+
+const initialState: userState = {
+  profile: null,
+  graph: null,
+};
+
+export const userSlice = createSlice({
+  name: "user",
+  initialState,
+  reducers: {
+    userLogin: (state, action: PayloadAction<userState>) => {
+      const profile: Profile = action.payload.profile as Profile;
+      const graph: Graph = action.payload.graph as Graph;
```

Comment:
The userState type is `type | null` since the user can be logged out and we don't know their profile or graph. If there's a better way to do this I can change it.

Reply:
Ah yes there is. So in this case if null were to come in, it would just get overlooked until it threw an NPE.

In this case you never expect that to happen, so the first option might be to check for null and throw if they are. That will fix the types afterward as TS is smart enough to infer that the types are no longer nullable:

```suggestion
      const profile = action.payload.profile;
      const graph = action.payload.graph;
      if (profile === null || graph === null) throw new Error("Login Failure");
```

(Note that error is just an example one).

Another way would be to create a version of `userState` that didn't allow for nulls in the first place. That way when you reduce `userLogin`, the parameters cannot be null. There are two utility types around this, but neither of them quite work. One is `NonNullable`, but it is for the type only, not the nested interface. Another utility type for it is https://www.typescriptlang.org/docs/handbook/utility-types.html#requiredtype but it only works for optionals instead of nullable. So an option would be to swap from null to optionals:

```suggestion

Interface UserState {
  profile?: Profile;
  graph?: Graph;
}

const initialState: UserState = {};

    userLogin: (state, action: PayloadAction<Required<UserState>>) => {
      const profile = action.payload.profile;
      const graph = action.payload.graph;
```

Also side node: types tend to be PascalCase.

Reply:
I went with optionals since it looks clean and readable plus throwing an error in a reducer felt bad.